File: code/python/proportion_with_one_cause.py
import dqaf_metrics

# Define metric
def metric(records):
    # Find all the Record Axis COD columns
    # Jurisdictions: update the column name prefix from "Record Axis COD" to match your data
    record_axis_cod_columns = [col for col in records.columns if col.startswith("Record Axis COD")]
    # Return a column that is True when a single Record Axis COD column is populated
    return records[record_axis_cod_columns].notna().sum(axis=1) == 1

description = "The proportion of records with a single Record Axis COD"

# Use metric to calculate overall proportion 
proportion = dqaf_metrics.proportion(metric)
print(f"{description} is {proportion:.2f}")

# Use metric to calculate proportion for each certifier
proportion_by_certifier = dqaf_metrics.proportion_by_column(metric, "Certifier Name")
# TODO: Update to use tabular output
for certifier, proportion in proportion_by_certifier:
    print(f"{description} provided by {certifier} is {proportion:.2f}")

File: code/R/proportion_with_one_cause.R
# Load libraries and supporting functions ----

# Load necessary libraries
library(here)

# Load supporting functions
source(here::here("code", "R", "dqaf_metrics.R"))

# Load variables common across all scripts -- to edit, see dqaf_common_variables.R
source(here::here("code", "R", "dqaf_common_variables.R"))

# Specify data and column names ----

# Load the death records data
file_path <- here::here("data", "SyntheticDeathRecordData.csv")
death_records <- 
  load_death_records(file_path)

# Specify required metric column names here
cause_of_death_columns <- c(
  colnames(death_records)[startsWith(colnames(death_records), "Record Axis COD")]
)

# Specify certifier column name here
certifier_name_column <- "Certifier Name"

# Calculate metric ----

# Create a new column that is True when a single Record Axis COD column is populated
death_records[, "Single Record Axis COD"] <-
  rowSums(!is.na(
    death_records[, cause_of_death_columns, drop = F])
  ) == 1

# Calculate the proportion of records with only one Record Axis COD
proportion <- calculate_proportion(
  death_records, 
  metric = "Single Record Axis COD",
  metric_description = "Single Record Axis COD", 
  print_output = TRUE
)

# Group the records by certifier and calculate the proportion of flagged records for each certifier
certifier_proportions <- calculate_proportion_by_column(
  death_records, 
  metric = "Single Record Axis COD",
  column = certifier_name_column, 
  print_output = TRUE
)
